router/core/router.go

```diff
@@ -493,6 +497,37 @@ func (r *Router) bootstrap(ctx context.Context) error {
 	return nil
 }
 
+func (r *Router) authenticationMiddleware(ctx context.Context) func(http.Handler) http.Handler {
+	if len(r.authenticators) == 0 && !r.authenticationRequired {
+		return nil
+	}
+	authenticators := r.authenticators
+	if len(authenticators) == 0 {
+		// r.authenticationRequired must be true at this point
+		r.logger.Warn("authentication is required but no authenticators are configured")
+	}
+	authenticationRequired := r.authenticationRequired
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			auth, err := authentication.AuthenticateHTTPRequest(ctx, authenticators, r)
+			if err != nil {
+				w.WriteHeader(http.StatusForbidden)
+				_, _ = io.WriteString(w, http.StatusText(http.StatusForbidden))
+				return
+			}
+			if auth != nil {
+				r = r.WithContext(authentication.NewContext(r.Context(), auth))
+				w.Header().Set("X-Authenticated-By", auth.Authenticator())
+			} else if authenticationRequired {
+				w.WriteHeader(http.StatusUnauthorized)
```

Comment:
This needs to be aligned with https://github.com/wundergraph/cosmo/blob/main/router/core/graphql_handler.go#L301 to track errros and respond in proper GraphQL response. Any particular reason, why you don't implement it as a custom module inside `Middleware(ctx RequestContext, next http.Handler)` ? In that way, you can add this module at server startup.

Reply:
Smiliar as in https://github.com/wundergraph/cosmo/blob/main/router/core/header_rule_engine.go.

Reply:
As discussed, this has been refactored to play nicely with metrics and tracing